### ISSUE TITLE
CV11 swaps the arguments of MySQL CONVERT(expr, type)

### ISSUE THREAD

**Issue by chuenchen309** [sqlfluff/sqlfluff#8171]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues. (The existing MySQL `CONVERT` issues — #6277, #6255 — are about `ALTER TABLE ... CONVERT TO CHARACTER SET`, which is unrelated.)

### What Happened

MySQL's `CONVERT` takes its arguments the other way round from T-SQL's: `CONVERT(expr, type)` rather than `CONVERT(type, expr)`. CV11 assumes the T-SQL order for every dialect, so on `mysql` it rewrites the two arguments into a `CAST` the wrong way round — silently, because the result still parses.

I'm opening an issue rather than a PR because the right fix looks like a judgement call, and I'd rather ask than guess. Details at the bottom.

### Expected Behaviour

```sql
SELECT CONVERT(b, SIGNED) FROM t
```

should either be rewritten to `CAST(b AS SIGNED)`, or left alone — but not have its arguments swapped.

### Observed Behaviour

`sqlfluff fix` produces a cast of the *type name* to the *column*:

```sql
SELECT cast(SIGNED as b) FROM t
```

`SIGNED` is now the value being cast and `b` is the target type. `DECIMAL(10,2)` goes the same way:

```
CONVERT(b, DECIMAL(10,2))  ->  cast(DECIMAL(10,2) as b)
```

For contrast, the same rule on `tsql` — where the argument order it assumes is the real one — is correct:

```
tsql:  CONVERT(INT, b)  ->  cast(b as INT)     # right
mysql: CONVERT(b, SIGNED)  ->  cast(SIGNED as b)  # wrong
```

`CONVERT(b USING utf8mb4)` is correctly left alone.

### How to reproduce

`.sqlfluff`:

```ini
[sqlfluff]
dialect = mysql
rules = CV11

[sqlfluff:rules:convention.casting_style]
preferred_type_casting_style = cast
```

`a.sql`:

```sql
SELECT CONVERT(b, SIGNED) FROM t
```

```
$ sqlfluff fix --rules CV11 -f a.sql
$ cat a.sql
SELECT cast(SIGNED as b) FROM t
```

### Dialect

`mysql`, and the dialects that inherit it: `mariadb`, `doris`, `starrocks`.

### Version

```
$ sqlfluff --version
python -m sqlfluff, version 4.2.2
```
Reproduced on `main` at e92448c.

---

### Why I'm asking instead of sending a patch

While digging into this I found the surrounding picture is a bit more tangled than "CV11 has the argument order wrong", and I don't want to pick a direction on your behalf:

**Most MySQL `CONVERT` forms don't parse at all today.** `CONVERT(b, CHAR)` and `CONVERT(price, UNSIGNED)` are unparsable in the `mysql` dialect:

```
L:   1 | P:  15 |  PRS | Line 1, Position 15: Found unparsable section: '(b, CHAR)'
```

They're untouched by CV11 not because anything guards them, but because parsing already failed. `CONVERT(b, SIGNED)` happens to parse — `SIGNED` isn't a reserved keyword, so it comes through as a plain `column_reference` — and that's the case that gets mangled. `grep -i convert dialect_mysql.py` only turns up `ALTER ... CONVERT TO CHARACTER SET`, so as far as I can tell the `mysql` dialect doesn't model `CONVERT(expr, type)` at all.

So there seem to be two ways to go, and they're quite different in scope:

1. **Teach the `mysql` dialect about `CONVERT(expr, type)`**, which would fix the unparsable forms too and let CV11 rewrite it correctly.
2. **Have CV11 decline to rewrite `CONVERT` on the MySQL-family dialects.** This stops the corruption but leaves the unparsable forms as they are.

For what it's worth, (2) has a precedent sitting in the same function — the `cast_expression` branch already refuses to touch dialect-specific variants for exactly this reasoning:

> CV11 can only rewrite the standard `::` shorthand. Dialect-specific variants such as Databricks `?::` (try_cast) and Vertica `::!` (null cast) are not interchangeable with CAST/CONVERT, so rewriting them would change semantics (and may even be unparsable).

MySQL's `CONVERT` looks like the same category to me. But (1) is clearly the more complete answer, and you'd know better than I would whether it's wanted.

Happy to put up a PR for whichever you'd prefer.

`CV11.yml` currently has no `mysql`/`mariadb` cases, which is presumably why this hasn't surfaced.

### Disclosure

AI-assisted, and I'd rather say so plainly. Everything above — the reproducer, the tsql contrast, the unparsable forms, the dialect grep — I ran and read myself rather than assumed.

**Comment by peterbud** [comment]
Hi @chuenchen309  - I'd go first with making MySQL dialect parse 'CONVERT' properly

**Comment by chuenchen309** [comment]
Thanks — that's a clearer answer than I expected, and it makes the CV11 half mostly moot: once the dialect gives the rule real nodes instead of two bare `column_reference`s, there's something honest to branch on.

I went and mapped what the `mysql` dialect does with `CONVERT` today, since "it doesn't parse" turned out to undersell it. Every form is wrong, in one of two ways:

**Unparsable** — the type name is a reserved word, so the argument grammar rejects it:

```
CONVERT(b, CHAR)            Found unparsable section: '(b, CHAR)'
CONVERT(b, UNSIGNED)        "
CONVERT(b, SIGNED INTEGER)  "
CONVERT(b, BINARY)          "
CONVERT(b, DOUBLE)          "
CONVERT(b, FLOAT)           "
CONVERT(b USING utf8mb4)    "
```

**Parses into the wrong shape** — the type isn't reserved, so it comes through as a value:

```
CONVERT(b, DATE)           -> column_reference 'b', column_reference 'DATE'
CONVERT(b, JSON)           -> column_reference 'b', column_reference 'JSON'
CONVERT(b, SIGNED)         -> column_reference 'b', column_reference 'SIGNED'
CONVERT(b, DECIMAL(10,2))  -> column_reference 'b', function DECIMAL(10, 2)
CONVERT(b, CHAR(10))       -> column_reference 'b', function CHAR(10)
```

That second group is the one that bites: CV11 sees a two-argument `CONVERT` function and applies the T-SQL reading, which is where `cast(SIGNED as b)` comes from. The first group is only "safe" because parsing already failed.

So the plan is a `CONVERT` grammar in `dialect_mysql.py` covering `CONVERT(expr, type)` over the documented cast types (`SIGNED`/`UNSIGNED` [`INTEGER`], `BINARY[(n)]`, `CHAR[(n)]`, `NCHAR[(n)]`, `DATE`, `DATETIME[(n)]`, `TIME[(n)]`, `DECIMAL[(m[,d])]`, `DOUBLE`, `FLOAT[(p)]`, `YEAR`, `JSON`) plus `CONVERT(expr USING charset)`, with dialect fixtures. `mariadb`/`doris`/`starrocks` inherit from `mysql`, so they come along.

Shout if you'd scope it differently — in particular whether you want `CAST(... AS ...)`'s MySQL-specific types in the same pass or separately. I'll leave CV11 untouched until the grammar lands, then come back to it.

One practical note: I'm at my self-imposed limit for PRs opened against sqlfluff today, so this will land tomorrow rather than in the next hour.

**Comment by adhavan18** [comment]
@chuenchen309 are you still on this one? No rush and no claim intended, your mapping above is more complete than anything I would have started from, and the plan you sketched looks right to me.

Asking because it has been a few weeks and the `CONVERT` grammar has not landed yet, and this is a silent-corruption bug rather than a cosmetic one: `sqlfluff fix` currently turns `CONVERT(b, SIGNED)` into `cast(SIGNED as b)` on mysql, which still parses, so a user can have their SQL quietly rewritten wrong.

If you have moved on, I am happy to pick up the dialect half exactly as you scoped it, `CONVERT(expr, type)` over the documented cast types plus `CONVERT(expr USING charset)`, with dialect fixtures, and to credit this analysis in the PR. If you are still on it, I will stay out of the way.

One thing I would suggest either way, since it is cheap and independent: CV11 could skip `mysql` outright until the grammar lands. Right now the rule fires on a shape it fundamentally misreads, so a two-line dialect guard would stop the bad autofix today without waiting on the parser work. Happy to send that separately if @peterbud wants it as a stopgap.

**Comment by adhavan18** [comment]
Opened #8283 with the CV11 guard I mentioned, so `fix` stops corrupting SQL while the dialect work is decided. It does not touch the parser, so it should not collide with anything you do here.

One thing that turned up while testing it, which is not in the issue: **the corruption runs in both directions.** I was checking `CONVERT` to `CAST` and wrote a fixture for the reverse, expecting it to pass. It failed. With `preferred_type_casting_style = convert` on mysql:

```sql
SELECT CAST(b AS SIGNED) FROM t   ->   SELECT convert(SIGNED, b) FROM t
```

Same T-SQL argument order, same silent breakage, just on the way out instead of the way in. So the rule cannot safely emit `CONVERT` on these dialects either, and both directions are skipped in the PR. Worth knowing for the grammar work: once the dialect parses `CONVERT` properly, `_convert_fix_list` will need to build the mysql order as well as read it.

@chuenchen309 still no rush on the dialect half from me, and the guard is written to be removed once the rule can tell the orders apart.

**Comment by chuenchen309** [comment]
Straight answer: **no, I'm not on it — please take the dialect half.** I said "tomorrow" on Jul 17 and then didn't deliver for three weeks, which is exactly the situation where someone else should not be waiting on me. Take the scope as written, no credit needed beyond whatever you think is fair.

Your stopgap was the right call, and I should have made it myself instead of leaving the rule firing on a shape it misreads while the parser work sat unstarted.

**Your reverse-direction finding checks out** — I went and read it rather than taking it on trust, and it's the same assumption baked in twice. In `CV11._eval`, the `CAST` → `CONVERT` branch does:

```python
cast_content = self._get_children(bracketed)
fixes = self._convert_fix_list(
    context,
    cast_content[1],   # data_type
    cast_content[0],   # expression
)
```

and `_convert_fix_list` emits `convert(arg1, arg2)` — so the type is placed first, T-SQL order, giving `convert(SIGNED, b)` on mysql just as you saw. The inbound branch is the mirror image, `_cast_fix_list(context, [convert_content[1]], convert_content[0])`, reading argument 0 as the type. Neither branch has any dialect awareness, so skipping both on mysql is the only correct guard short of the grammar landing.

Worth noting for whoever does the grammar work: this means `_convert_fix_list` can't simply be reused for mysql later. Once the dialect parses `CONVERT(expr, type)`, the emit path needs a dialect-conditional argument order, not just a read path that understands both — otherwise `preferred_type_casting_style = convert` will keep producing T-SQL order on a dialect that has just learned to reject it.

One correction to my earlier comment while I'm here: I listed `CONVERT(b, DATE)` and friends as "parses into the wrong shape". That was measured against the rule's behaviour, which is what mattered for CV11, but for the grammar work the useful framing is the one you'd hit first — those forms parse into `column_reference` where a `data_type` belongs, so the fixtures need to assert on node type, not just on parseability.

@peterbud — @adhavan18 is picking up the `CONVERT` grammar for mysql on the scope sketched above; I'm stepping back from it.
